// @nomicfoundation/hardhat-ledger/dist/src/internal/cache.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.read = exports.write = exports.CACHE_FILE_NAME = void 0;
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
exports.CACHE_FILE_NAME = "accounts.json";
const PACKAGE_NAME = "hardhat";
async function write(json) {
    const ledgerCacheFile = await getLedgerCacheFile(exports.CACHE_FILE_NAME);
    await fs_extra_1.default.writeJSON(ledgerCacheFile, json);
}
exports.write = write;
async function read() {
    const ledgerCacheFile = await getLedgerCacheFile(exports.CACHE_FILE_NAME);
    try {
        const file = await fs_extra_1.default.readJSON(ledgerCacheFile);
        return file;
    }
    catch (error) { }
}
exports.read = read;
async function getLedgerCacheFile(fileName) {
    const ledgerCacheDir = await getLedgerCacheDir();
    return path_1.default.join(ledgerCacheDir, fileName);
}
async function getLedgerCacheDir() {
    const cache = await getCacheDir();
    const compilersCache = path_1.default.join(cache, "ledger");
    await fs_extra_1.default.ensureDir(compilersCache);
    return compilersCache;
}
async function getCacheDir() {
    const { cache } = await generatePaths(PACKAGE_NAME);
    await fs_extra_1.default.ensureDir(cache);
    return cache;
}
async function generatePaths(packageName) {
    const { default: envPaths } = await Promise.resolve().then(() => __importStar(require("env-paths")));
    return envPaths(packageName);
}
